src/include/rbd/librbd.hpp

```diff
@@ -356,6 +356,8 @@ class CEPH_RBD_API RBD
 
   int mirror_mode_get(IoCtx& io_ctx, rbd_mirror_mode_t *mirror_mode);
   int mirror_mode_set(IoCtx& io_ctx, rbd_mirror_mode_t mirror_mode);
+  int mirror_mode_set2(IoCtx& io_ctx, rbd_mirror_mode_t mirror_mode,
+                       const std::string &ns_name);
```

Comment:
>  [...] initial review comments, specifically around the API changes.

I know we discussed enforcement at the API level and having it be a single call like this is one way to do it, but after looking at the code having an API that deals with the mode also deal with the remote namespace doesn't seem entirely right to me.  A corresponding getter is needed which would likely be asymmetric and because `Mirror::mode_set()` is also used for disabling mirroring a dummy string needs to be passed at that point just to satisfy the signature.

I don't think there is any requirement on `mirror_mode_set()` being the thing that is called first and creates `rbd_mirroring` object.  How about introducing a separate pair of `mirror_remote_namespace_get()` and `mirror_remote_namespace_set()` APIs with the assumption that `mirror_remote_namespace_set()` would be called before `mirror_mode_set()`?  This could be enforced with a "if `MODE` key exists, proceed only if value is `MIRROR_MODE_DISABLED`" check in `mirror_remote_namespace_set` CLS method, which is really nice because this way it would be _actually_ atomic.
```
  int mirror_remote_namespace_get(IoCtx& io_ctx, std::string* remote_namespace);
  int mirror_remote_namespace_set(IoCtx& io_ctx, const std::string& remote_namespace);
```

Reply:
>  A corresponding getter is needed

... and `rbd mirror pool info` command showing `Remote namespace` after `Mode` when run on namespaces.

We also need the corresponding C APIs and Python bindings.

Reply:
> > A corresponding getter is needed
> 
> ... and `rbd mirror pool info` command showing `Remote namespace` after `Mode` when run on namespaces.
> 
> We also need the corresponding C APIs and Python bindings.

These are already listed in the doc .

Reply:
> This could be enforced with a "if `MODE` key exists, proceed only if value is `MIRROR_MODE_DISABLED`" check in `mirror_remote_namespace_set` CLS method, which is really nice because this way it would be _actually_ atomic.

I will add the additional APIs but I think the MODE check would be better at the API level itself. That way the cls methods only handle what they need to (get and set the remote ns key).

Reply:
Actually, having these APIs will cause issues if they can be directly called without having the cls enforce the check. How about a single mirror_pool_enable api instead which will take the remote ns as well and perform all these checks internally? It could attempt to set the remote ns if it present and then call the existing Mirror::mode_set method.

Reply:
How will this feature work on mixed clusters? Is the following possible?

On a node running the new rbd cli, enable mirroring on namespace with the --remote-namespace option. 

If the rbd-mirror daemon is running on a node with an older version, it will not use the remote_namespace.

If the ns is then disabled and enabled without the --remote-namespace option on a node running an older version of ceph, the remote_namespace key will still exist in the rbd_mirroring object and be picked up by the NamespaceReplayer if the rbd-mirror daemon is running on a node which has remote-namespace support.

Reply:
> I will add the additional APIs but I think the MODE check would be better at the API level itself.

The check wouldn't be atomic if done at the API level -- consider a case of two `rbd mirror pool enable` commands racing with each other, for example.  Granted, scenarios such as this is something that we don't always handle -- in some cases, it's an implicit "don't do this, if you do you will get in trouble" sort of handling. But in this case, making the CLS method take care of the potential race is trivial.

> That way the cls methods only handle what they need to (get and set the remote ns key).

The reason we bother with CLS methods instead of simply getting or setting the value using the public librados API is precisely to be able to do checks like this and/or enforce invariants atomically.  Some of the CLS methods still end up being pure getters and setters, but there is definitely no expectation that a CLS method concerns itself with just one key.

Reply:
> How will this feature work on mixed clusters? Is the following possible? [...]

Yes, I think all these scenarios are possible.  The only way to avoid them is to make an incompatible on-disk change to prevent an old rbd-mirror daemon from bootstrapping on namespaces where a custom remote namespace is configured (and possibly also prevent the mode from being manipulated altogether).  I don't see an easy way to do that though...

Since the worst that can happen is rbd-mirror daemon mirroring into the wrong namespace and namespace mirrorring even as it is today (i.e. without support for namespace remapping) is unlikely to be widely used, I'm inclined to just let this go.  It's probably worth a deeper discussion though.

Reply:
> The only way to avoid them is to make an incompatible on-disk change

There is also a theoretical option of tying it to a (required) cluster-wide feature bit, but it's not something that we have ever done for RBD and I doubt it would be feasible/cover all scenarios here even if a suitable bit was available.